### PR TITLE
feat: implement support for logrotate

### PR DESCRIPTION
**Pull request by s3lph**
When daliserver writes a logfile, the logfile is never rotated and keeps on growing. On our system it has now reached ~6GB. This ended up filling the disk of our backup server, as we do daily backups with file-level deduplication: As the file was constantly appended to, it was never the same file, and every incremental backup contained a full copy of this file.

While trying to set up a logrotate config, I realized that daliserver did not yet contain any logic for re-opening the logfile, which is requried for logrotate to work. So this PR contains two things:
- A logrotate config for the Debian package that signals the `daliserver` process with `SIGHUP` after log rotation
- A signal handler for `SIGHUP` that re-opens the logfile by name.

### CODE REVIEW

File: lib/log.c [hunk 1]
```diff
@@ -155,7 +156,11 @@ unsigned int log_get_level() {
 	return loglevel;
 }
 
-int log_set_logfile(const char *logfile) {
+int log_reopen_file() {
+	if (fp_logfile) {
+		fclose(fp_logfile);
```
Comment (onitake) [inline lib/log.c:161]: I'd prefer to use `freopen()` instead of closing and reopening the log file.
That way, there's no need to hold the log file name in a global variable.

File: lib/dispatch.c [hunk 2]
```diff
@@ -73,7 +73,7 @@ DispatchStatus dispatch_run(DispatchPtr table, int timeout) {
 			if (errno != EINTR) {
 				log_error("Error waiting for I/O events");
 			}
-			return 0;
+			return 1;
```
Comment (onitake) [inline lib/dispatch.c:76]: Is this a fix for a bug (i.e. an incorrect return value), or what's the purpose of this change? :thinking:

**Review by onitake — CHANGES_REQUESTED** [review]
Thanks for the PR!

Aside from my comments, I'd like to point out that daliserver also prints fatal and error messages to STDERR and all other messages to STDOUT.
If you're running daliserver in something that captures STDOUT/STDERR (such as systemd+journald), there's no need to enable the log file or syslog at all. I'd strongly recommend to use this instead.

File: lib/log.c (hunk 1 above)
Comment (s3lph) [reply lib/log.c:161]: `freopen` too takes a filename. And the file needs to be reopened by name, as after a logrotate there's an entirely new file under the name. 

`freopen` with a pathname set to NULL merely changes the mode of the already open file, but will still refer to the same file, which will by then have moved to `daliserver.log.1`.

File: lib/dispatch.c (hunk 2 above)
Comment (s3lph) [reply lib/dispatch.c:76]: `poll` will be interrupted by the SIGHUP as well, and returning `0` here causes the main program to break out of the event loop and terminate:

https://github.com/onitake/daliserver/blob/4cfdb984e9e054d9e9008815d4253c9ba551fef3/src/daliserver.c#L180

If instead, `1` is returned here, the same loop in the main program will consider the value of `running` as well, which will still be `1` after a SIGHUP, but `0` after a SIGINT/SIGTERM.

So yes, with the introduced distinction between SIGHUP and SIGINT/SIGTERM, returning `0` here would be a bug.

**Comment by s3lph** [comment]
> I'd like to point out that daliserver also prints fatal and error messages to STDERR and all other messages to STDOUT. If you're running daliserver in something that captures STDOUT/STDERR (such as systemd+journald), there's no need to enable the log file or syslog at all. I'd strongly recommend to use this instead.

Nevertheless, the default behavior of the `daliserver` Debian package still is to use an init script, which starts dalisever with `-f /var/log/daliserver.log`. As long as this is the default behavior, it should also support log rotation.

I do however also agree that systemd+journald would be the better way forward, but that's a change I'd like to see in the daliserver Debian package itself, rather than manually installing a custom systemd unit alongside the package.

File: lib/log.c (hunk 1 above)
Comment (onitake) [reply lib/log.c:161]: It's unfortunate that the [POSIX specification](https://pubs.opengroup.org/onlinepubs/9699919799/functions/freopen.html) is so vague in terms of how `freopen(NULL,...)` should be handled.

glibc actually _does_ reopen and close the file. However, it does so by opening `/proc/self/fd/<original file descriptor>`, which evidently causes the Linux kernel to merely duplicate the file kernel handle plus the changed attributes. If the directory entry has changed in the meantime, this will not be reflected by the new handle. Fun fact: The kernel creates a new file descriptor, but freopen calls dup2() to reuse the old one.

Apparently, some other UNIXes don't even support this at all.

**Comment by onitake** [comment]
> Nevertheless, the default behavior of the `daliserver` Debian package still is to use an init script, which starts dalisever with `-f /var/log/daliserver.log`. As long as this is the default behavior, it should also support log rotation.
> 
> I do however also agree that systemd+journald would be the better way forward, but that's a change I'd like to see in the daliserver Debian package itself, rather than manually installing a custom systemd unit alongside the package.

It's high time I make that switch, then. :slightly_smiling_face: 

Sorry, dear Devuan users...

File: lib/log.c [hunk 3]
```diff
@@ -167,13 +172,21 @@ int log_set_logfile(const char *logfile) {
...
+}
+
+int log_set_logfile(const char *logfile_path) {
+	if (logfile_path == NULL) {
+		logfile = NULL;
 	} else {
-		if (fp_logfile) {
-			fclose(fp_logfile);
-			fp_logfile = NULL;
+		logfile = strdup(logfile_path);
```
Comment (onitake) [inline lib/log.c]: This will leak memory if the log file name is changed during program execution.
It's not a problem now, but I'd prefer that we guard against that.

**Review by onitake — CHANGES_REQUESTED** [review]

**Comment by s3lph** [comment]
> Sorry, dear Devuan users...

If you would want to make Devuan users happy, you could ship both `/etc/init.d/daliserver` and `/lib/systemd/system/daliserver.service`. Systemd won't use sysv-generator to generate a unit if one with the same name already exists. So if running under systemd, the native unit is used, whereas if running under an init system that uses `/etc/init.d` scripts the script is used.

File: lib/log.c (hunk 3 above)
Comment (s3lph) [reply lib/log.c]: Added a `free`, please have another look.

**Comment by onitake** [comment]
> If you would want to make Devuan users happy, you could ship both `/etc/init.d/daliserver` and `/lib/systemd/system/daliserver.service`. Systemd won't use sysv-generator to generate a unit if one with the same name already exists. So if running under systemd, the native unit is used, whereas if running under an init system that uses `/etc/init.d` scripts the script is used.

Please take a look: #28

File: lib/log.c (hunk 3 above)
Comment (onitake) [reply lib/log.c]: That part is good, thanks.

**Review by onitake — CHANGES_REQUESTED** [review]
With the changed logic, part of https://github.com/onitake/daliserver/pull/27/files#diff-f7dfb3ca741707d1a361d8cdd6f75f05bc99811daac01afaa258f43d4b71e4f6R165-R174 is now redundant and can be simplified. Please remove the extra fclose(), and fold the assignment to fp_logfile into the fopen() line. There's no need for an extra temp var.

**Comment by s3lph** [comment]
> Please remove the extra fclose(), and fold the assignment to fp_logfile into the fopen() line. There's no need for an extra temp var.

Done.

**Review by onitake — APPROVED** [review]
That looks good, thank you.
